### PR TITLE
vendor, trie: replace bigCache with fastCache

### PR DESCRIPTION
**Pull request by rjl493456442**
This PR replaces the bigCache with fastCache. 

FastCache is mainly for reducing the memory usage which is really nice for us. 
Besides, fastCache can also support cache dump and reload, so that we 
don't need the warm time to re-fill the cache when we restart.

### CODE REVIEW

**Comment by holiman** [comment]
I once submitted a fuzzer to bigcache, perhaps we could adapt it to fuzz fastCache also. It found a couple of issues that got fixed.Check my PRs to bigcache (I'm on mobile now, can check myself later)

**Comment by rjl493456442** [comment]
@holiman Thanks for the information, I think it's https://github.com/allegro/bigcache/pull/120/files. I can do a similar thing for fast-cache.

File: trie/database.go [hunk 1]
```diff
@@ -413,7 +426,9 @@ func (db *Database) Node(hash common.Hash) ([]byte, error) {
 	}
 	// Retrieve the node from the clean cache if available
 	if db.cleans != nil {
-		if enc := db.cleans.Get(nil, hash[:]); enc != nil {
+		buf := bufferPool.Get().([]byte)
+		defer bufferPool.Put(buf)
+		if enc := db.cleans.Get(buf[:0], hash[:]); len(enc) != 0 {
```
Comment (karalabe) [inline trie/database.go]: `enc` is essentially `buf` and it's returned to the caller, so we can't put it back into the sync pool

**Comment by rjl493456442** [comment]
Add a few benchmark result here

**mon07 represents fastCache, mon06 represents bigCache**

### Memory allocation

image

**Memory allocation wise, bigCache allocates more ~800MB memory**

### Memory usage

image

**similarly, bigCache uses more ~800MB memory**

### Performance hit

image

**Performance-wise, basically there is no big difference**

**Comment by karalabe** [comment]
We need to figure out how to do vendoring / building properly with this PR. We need to ensure that building from vendor works and that also building with go-ethereum as a dependency (without vendor) works. We might need to push the go modules rework.

**Comment by karalabe** [comment]
Screenshot from 2019-08-23 16-00-36

Screenshot from 2019-08-23 16-00-49

**Comment by rjl493456442** [comment]
There is a problem to integrate `FastCache` into `Geth` since in `FastCache` it uses v2 `xxhash` based on the Module mode.  While for go-ethereum, it still uses vendor mode to manage all dependencies. 

Check [how-are-v2-modules-treated-in-a-build-if-modules-support-is-not-enabled-how-does-minimal-module-compatibility-work](https://github.com/golang/go/wiki/Modules#how-are-v2-modules-treated-in-a-build-if-modules-support-is-not-enabled-how-does-minimal-module-compatibility-work-in-197-1103-and-111) 

Golang team introduces an idea of "minimal-module-compatibility". In short, in this case, go tool will try to find package under the `GOPATH` without the `vn` 

> For example, when operating in GOPATH mode, import "foo/v2" appearing in module-based code resolves to the same code residing in your GOPATH as import "foo", and the build ends up with one copy of foo – in particular, whatever version is on disk in GOPATH. This allows module-based code with import "foo/v2" to compile even in GOPATH mode in 1.9.7+, 1.10.3+ and 1.11.

But the go tool will only find package under the `GOPATH` but not `vendor` dir under the project. 
Not sure it's supported but seems not.

**Comment by holiman** [comment]
Fuzzed fastcache a bit, looks good to me

**Pull request merged**